ultralist/date_parser.go

```diff
@@ -0,0 +1,111 @@
+package ultralist
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+// DateParser is a thing that parses relative, arbitrary, and absolute dates, and returns a date in the format of yyyy-mm-dd
+type DateParser struct{}
+
+// ParseDate takes a date from a Filter and turns it into a string with the format of yyyy-mm-dd
+func (dp *DateParser) ParseDate(dateString string, pivotDay time.Time) (date time.Time, err error) {
+	switch dateString {
+	case "none":
+		return time.Time{}, nil
+	case "yesterday", "yes":
+		return bod(pivotDay).AddDate(0, 0, -1), nil
+	case "today", "tod":
+		return bod(pivotDay), nil
+	case "tomorrow", "tom":
+		return bod(pivotDay).AddDate(0, 0, 1), nil
+	case "monday", "mon":
+		return dp.monday(pivotDay), nil
+	case "tuesday", "tue":
+		return dp.tuesday(pivotDay), nil
+	case "wednesday", "wed":
+		return dp.wednesday(pivotDay), nil
+	case "thursday", "thu":
+		return dp.thursday(pivotDay), nil
+	case "friday", "fri":
+		return dp.friday(pivotDay), nil
+	case "saturday", "sat":
+		return dp.saturday(pivotDay), nil
+	case "sunday", "sun":
+		return dp.sunday(pivotDay), nil
+	case "lastweek":
+		n := bod(pivotDay)
+		return getNearestMonday(n).AddDate(0, 0, -7), nil
+	case "nextweek":
+		n := bod(pivotDay)
+		return getNearestMonday(n).AddDate(0, 0, 7), nil
+	}
+	return dp.parseSpecificDate(dateString, pivotDay)
+}
+
+func (dp *DateParser) monday(day time.Time) time.Time {
+	mon := getNearestMonday(day)
+	return dp.thisOrNextWeek(mon, day)
+}
+
+func (dp *DateParser) tuesday(day time.Time) time.Time {
+	tue := getNearestMonday(day).AddDate(0, 0, 1)
+	return dp.thisOrNextWeek(tue, day)
+}
+
+func (dp *DateParser) wednesday(day time.Time) time.Time {
+	tue := getNearestMonday(day).AddDate(0, 0, 2)
+	return dp.thisOrNextWeek(tue, day)
+}
+
+func (dp *DateParser) thursday(day time.Time) time.Time {
+	tue := getNearestMonday(day).AddDate(0, 0, 3)
+	return dp.thisOrNextWeek(tue, day)
+}
+
+func (dp *DateParser) friday(day time.Time) time.Time {
+	tue := getNearestMonday(day).AddDate(0, 0, 4)
+	return dp.thisOrNextWeek(tue, day)
+}
+
+func (dp *DateParser) saturday(day time.Time) time.Time {
+	tue := getNearestMonday(day).AddDate(0, 0, 5)
+	return dp.thisOrNextWeek(tue, day)
+}
+
+func (dp *DateParser) sunday(day time.Time) time.Time {
+	tue := getNearestMonday(day).AddDate(0, 0, 6)
+	return dp.thisOrNextWeek(tue, day)
+}
+
+func (dp *DateParser) thisOrNextWeek(day time.Time, pivotDay time.Time) time.Time {
+	if day.Before(pivotDay) {
+		return day.AddDate(0, 0, 7)
+	}
+	return day
+}
+
+func (dp *DateParser) parseSpecificDate(date string, pivot time.Time) (time.Time, error) {
```

Comment:
are we missing the obvious internal date format, 2006-01-02?

Reply:
parsing specific dates from the user would be in the format of `jan2` or `2jan`.